_posts/2023-02-07-homebrew-4.0.0.md

```diff
@@ -0,0 +1,53 @@
+---
+title: 4.0.0
+author: MikeMcQuaid
+redirect_from: /blog/4.0.0/
+---
+
+Today I'd like to announce Homebrew 4.0.0. The most significant change since 3.6.0 is faster updating through JSON downloads.
+
+Major changes and deprecations since 3.6.0:
+
+- Using JSON files downloaded from [formulae.brew.sh](https://formulae.brew.sh) rather than local homebrew-core and homebrew-cask taps.
+  - **Please note: this is the largest change we have made to our update process since we split Homebrew/brew and Homebrew/homebrew-core repositories. Please bear with us, there may be a few bumps**.
+  - Unless you are developing formulae or casks, you can `brew untap homebrew/core` and `brew untap homebrew/cask` to save some space.
+  - [`brew update` will now be run automatically less often (every 24 hours rather than every 5 minutes)](https://github.com/Homebrew/brew/pull/14376) and these auto-updates will be much faster as they no longer need to perform the slow `git fetch` of the huge homebrew-core and homebrew-cask taps' Git repositories.
```

Comment:
I'd love it if you could provide some concrete metric, at least something like "order of magnitude faster" or some factor like that, even if's entirely synthetic.

Reply:
Hesistant to do this without actually measuring and can't really get a concrete value for this given the variability.

Reply:
Assuming that network speed for a particular individual is the same when accessing Homebrew repositories, we can provide "back-of-the-envelope" calculations. Here is an example:

Previously, in order to use Homebrew suite on a macOS system, one had to download (and keep updating):

- **Homebrew/brew**
- **Homebrew/core**
- **Homebrew/cask** (on macOS)

After the change, users will have to download (and keep updating) **Homebrew/brew** only and then get tiny jsons.

Here is the breakdown of things:

**Homebrew/brew**: Initial (first-time) full clone: 40s, `.git/` directory: 72 MiB, Commits: 34,000
**Homebrew/core**: Initial (first-time) full clone: 7m 20s, `.git/` directory: 579 MiB, Commits: 346,000
**Homebrew/cask**: Initial (first-time) full clone: 4m, `.git/` directory: 352 MiB, Commits: 170,000

From the above we see that average "commit" sizes are:
- 2.1 KiB in Homebrew/brew
- 1.7 KiB in Homebrew/core
- 2.2 KiB in Homebrew/cask

Repository age:
- 5014 days for Homebrew/brew
- 5000 days for Homebrew/core (🎉, btw)
- 3722 days for Homebrew/cask

So, on average Homebrew repositories produced:
- 7 commits/day by Homebrew/brew
- 70 commits/day by Homebrew/core
- 46 commits/day by Homebrew/cask

Translating that to download size, Homebrew repositories required per day:
- 7 commits/day &times; 2.1 KiB/commit = 14.7 KiB/day
- 70 commits/day &times; 1.7 KiB/commit = 119 KiB/day
- 46 commits/day &times; 2.2 KiB/commit = 101 KiB/day

In summary:
- before the change users had to spend 235 KiB/day.
- after the change users will have to spend 14.7 KiB/day.

This will make things 15 times faster (235/15).